DESIGN.md

```diff
@@ -0,0 +1,756 @@
+});
+```
+
+## Buffer management
+
+Buffers passed to read and write operations must remain alive and pinned to a
+memory location while operations are in-flight, ruling out `&mut [u8]` as an
+option. Additionally, read and write operations may reference buffers using a
+pointer or, when the buffer is part of a pre-registered buffer pool, using a
+numeric buffer identifier. 
+
+The tokio-uring crate will provide its buffer type, `IoBuf`, to use when reading
+and writing.
+
+```rust
+pub struct IoBuf {
+    kind: Kind,
+}
+
+enum Kind {
+    /// A vector-backed buffer
+    Vec(Vec<u8>),
+
+    /// Buffer pool backed buffer. The pool is managed by io-uring.
+    Provided(ProvidedBuf),
+}
+```
+
+Internally, the IoBuf type is either backed by individually heap-allocated
+memory or a buffer pool entry. Acquiring an `IoBuf` is done via
+`IoBuf::with_capacity` or checking-out an entry from a pool. 
+
+```rust
+// Individually heap-allocated
+let my_buf = IoBuf::with_capacity(4096);
+
+// Checked-out from a pool
+match my_buffer_pool.checkout() {
+    Ok(io_buf) => ...,
+    Err(e) => panic!("buffer pool empty"),
+}
+```
+
+On drop, if the `IoBuf` is a buffer pool member, it is checked back in. If the
+kernel initially checked out the buffer as part of a read operation, an io-uring
+operation is issued to return it. Submitting the io-uring operation requires the
+buffer to remain on the same thread that checked it out and is enforced by
+making the `IoBuf` type `!Send`. 
+
+`IoBuf` provides an owned slice API allowing the caller to read to and write from
+a buffer's sub-ranges. 
+
+```rust
+pub struct Slice {
+    buf: IoBuf,
+    begin: usize,
+    end: usize,
+}
+
+impl IoBuf {
+    fn slice(self, range: impl ops::RangeBounds<usize>) -> Slice { .. }
+}
+
+// Write a sub-slice of a buffer
+my_file.write(my_io_buf.slice(10..20)).await
+```
+
+A slice end may go past the buffer's length but not past the capacity, enabling
+reads to uninitialized memory.
+
+```rust
+// The buffer's memory is uninitialized
+let mut buf = IoBuf::with_capacity(4096);
+let slice = buf.slice(0..100);
+
+assert_eq!(slice.len(), 0);
+assert_eq!(slice.capacity(), 100);
+
+// Read data from a file into the buffer
+let BufResult(res, slice) = my_file.read_at(slice, 0);
+
+assert_eq!(slice.len(), 100);
+assert_eq!(slice.capacity(), 100);
+```
+
+A trait argument for reading and writing may be possible as a future
+improvement. Consider a read API that takes `T: AsMut<[u8]>` for the buffer
+argument.
+
+```rust
+async fn read<T: AsMut<[u8]>>(&self, dst: T) { ... }
+
+struct MyUnsableBuf {
+    mem: [u8; 10],
+}
+
+impl AsMut<[u8]> {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.mem[..]
+    }
+}
+```
+
+This read function takes ownership of the buffer; however, any pointer to the
+buffer obtained from a value becomes invalid when the value moves. Storing the
+buffer value at a stable location while the operation is in-flight should be
+sufficient to satisfy safety.
```

Comment:
It's technically safe to return a new slice or pointer on each call to `as_mut()`, as bonkers as that may be, so it might need something like `unsafe trait TrustedAsMut: AsMut<[u8]> {}`

Reply:
If we hold ownership of T and guarantee to call as_mut only once, I think this is trusted.

Reply:
This is correct, but I am *pretty* sure that as long as `as_mut()` is called only once and is kept at a stable location, it should be fine.... but that is also why I am punting 😆